tests/integration/test_types_integration.py

```diff
@@ -142,6 +145,145 @@ def test_json(trino_connection):
+        .add_field(sql="TIMESTAMP '1582-10-05 01:23:45.123'", python=timestamp_during_switch) \
+        .add_field(sql="TIMESTAMP '1582-10-14 01:23:45.123'", python=timestamp_gregorian) \
+        .add_field(sql="TIMESTAMP '2001-08-22 01:23:45.123'", python=timestamp_3) \
+        .add_field(sql="TIMESTAMP '2001-08-22 01:23:45.123456'", python=timestamp_6) \
+        .add_field(sql="TIMESTAMP '2001-08-22 01:23:45.123900'", python=timestamp_round_up) \
+        .add_field(sql="TIMESTAMP '2001-08-22 01:23:45.123456111'", python=timestamp_6) \
+        .add_field(sql="TIMESTAMP '2001-08-22 01:23:45.123456789'", python=timestamp_round_down) \
+        .add_field(sql="TIMESTAMP '2001-08-22 01:23:45.123456111111'", python=timestamp_6) \
+        .add_field(sql="TIMESTAMP '2001-08-22 01:23:45.123456789123'", python=timestamp_round_down) \
+        .execute()
+
+
+def test_timestamp_with_timezone(trino_connection):
+    query_timestamp_with_timezone(trino_connection, '-08:00')
+    query_timestamp_with_timezone(trino_connection, '+08:00')
+    query_timestamp_with_timezone(trino_connection, '+05:30')
+    query_timestamp_with_timezone(trino_connection, 'US/Eastern')
+    query_timestamp_with_timezone(trino_connection, 'Asia/Kolkata')
+    query_timestamp_with_timezone(trino_connection, 'GMT')
+
+
+def query_timestamp_with_timezone(trino_connection, tz_str):
+    tz = create_timezone(tz_str)
+    timestamp_0 = datetime(2001, 8, 22, 11, 23, 45, 0).replace(tzinfo=tz)
+    timestamp_3 = datetime(2001, 8, 22, 11, 23, 45, 123000).replace(tzinfo=tz)
+    # Because of `PARAMETRIC_DATETIME` not being set, values are rounded to milliseconds
+    timestamp_6 = datetime(2001, 8, 22, 11, 23, 45, 123000).replace(tzinfo=tz)
+    timestamp_round_down = datetime(2001, 8, 22, 11, 23, 45, 123000).replace(tzinfo=tz)
+    timestamp_round_up = datetime(2001, 8, 22, 11, 23, 45, 124000).replace(tzinfo=tz)
+
+    SqlTest(trino_connection) \
+        .add_field(sql="CAST(null AS TIMESTAMP WITH TIME ZONE)", python=None) \
+        .add_field(sql=f"TIMESTAMP '2001-08-22 11:23:45 {tz_str}'", python=timestamp_0) \
+        .add_field(sql=f"TIMESTAMP '2001-08-22 11:23:45.123 {tz_str}'", python=timestamp_3) \
+        .add_field(sql=f"TIMESTAMP '2001-08-22 11:23:45.123 {tz_str}'", python=timestamp_3) \
+        .add_field(sql=f"TIMESTAMP '2001-08-22 11:23:45.123456 {tz_str}'", python=timestamp_6) \
+        .add_field(sql=f"TIMESTAMP '2001-08-22 11:23:45.123900 {tz_str}'", python=timestamp_round_up) \
+        .add_field(sql=f"TIMESTAMP '2001-08-22 11:23:45.123456111 {tz_str}'", python=timestamp_6) \
+        .add_field(sql=f"TIMESTAMP '2001-08-22 11:23:45.123456789 {tz_str}'", python=timestamp_round_down) \
+        .add_field(sql=f"TIMESTAMP '2001-08-22 11:23:45.123456111111 {tz_str}'", python=timestamp_6) \
+        .add_field(sql=f"TIMESTAMP '2001-08-22 11:23:45.123456789123 {tz_str}'", python=timestamp_round_down) \
```

Comment:
Another thing to test is something like `2001-08-22 11:23:45.000001999`, to make sure the rounding part handles the leading zeros.

Reply:
I will keep the fractional part as a `Decimal` with the supplied precision.

Reply:
Still something useful to test

Reply:
What's still missing is:

- tests for the leading 0
- tests for the overflow.